lci_strategic_plugin/src/lci_strategic_plugin.cpp

```diff
@@ -369,82 +369,242 @@ void LCIStrategicPlugin::handleGreenSignalScenario(const cav_srvs::PlanManeuvers
+      modified_remaining_time_upper = (modified_departure_speed_upper - starting_speed) / max_comfort_accel_;
 
-  if (starting_speed <= departure_speed)
-    modified_remaining_time = (sqrt(pow(starting_speed, 2) + (2 * max_comfort_accel_ * remaining_downtrack)) - starting_speed)/ max_comfort_accel_;
-  else
-    modified_remaining_time = (sqrt(pow(starting_speed, 2) + (2 * max_comfort_decel_ * remaining_downtrack)) - starting_speed) / max_comfort_decel_;
+      traj_upper.t1_ = current_time + modified_remaining_time_upper;
+      traj_upper.v1_ = modified_departure_speed_upper;
+      traj_upper.a1_ = max_comfort_accel_;
+      traj_upper.x1_ = traffic_light_downtrack;
 
-  if (starting_speed <= departure_speed)
-  {
-    params.a1_ = max_comfort_accel_;
-    params.v1_ = sqrt(pow(starting_speed, 2) + (2 * max_comfort_accel_ * remaining_downtrack));
+      traj_upper.t2_ = traj_upper.t1_;
+      traj_upper.v2_ = traj_upper.v1_;
+      traj_upper.a2_ = traj_upper.a1_;
+      traj_upper.x2_ = traj_upper.x1_;
+
+      traj_upper.modified_departure_speed = modified_departure_speed_upper;
+      traj_upper.modified_remaining_time = modified_remaining_time_upper;
+    }
+    else // NOTE: most likely will not happen as it would have happened at trajectory smoothing part
+    { 
+      ROS_DEBUG_STREAM("Failure case upper 2");
+      
+      double cruising_distance = remaining_downtrack - (pow(departure_speed, 2) - pow(starting_speed, 2))/ ( 2 * max_comfort_accel_);
+      if (cruising_distance < -EPSILON)
+      {
+        ROS_DEBUG_STREAM("Detected calculation failure in upper case 2");
+        calculation_success_upper = false;
+      }
+      modified_remaining_time_upper = (departure_speed - starting_speed) / max_comfort_accel_ + cruising_distance / departure_speed;
+
+      traj_upper.t1_ = current_time + (departure_speed - starting_speed) / max_comfort_accel_;
+      traj_upper.v1_ = departure_speed;
+      traj_upper.a1_ = max_comfort_accel_;
+      traj_upper.x1_ = starting_downtrack + (pow(departure_speed, 2) - pow(starting_speed, 2)) / (2 * max_comfort_accel_);
+
+      traj_upper.t2_ = current_time + modified_remaining_time_upper;
+      traj_upper.v2_ = departure_speed;
+      traj_upper.a2_ = 0;
+      traj_upper.x2_ = traffic_light_downtrack;
+
+      traj_upper.modified_departure_speed = departure_speed;
+      traj_upper.modified_remaining_time = modified_remaining_time_upper;
+    }
   }
-  else
+
+  if (departure_speed < starting_speed) // NOTE: most cases will not run due to departure speed being equal to speed limit
   {
-    params.a1_ = max_comfort_decel_;
-    params.v1_ = sqrt(pow(starting_speed, 2) + (2 * max_comfort_decel_ * remaining_downtrack));
-  }
+    if ((pow(departure_speed,2) - pow(starting_speed,2))/(2*max_comfort_decel_) >= remaining_downtrack)
+    {
+      ROS_DEBUG_STREAM("Failure case upper 3");
```

Comment:
It is not technically a failure case. It just means that the vehicle cannot reach to departure speed which is ok. You can say, HandleFailureCase -> Upper Trajectory -> Current Speed > Desired Departure Speed,  Actual Departure Speed > Desired Departure Speed